binary/src/Pos/Binary/Class/Core.hs

```diff
@@ -659,3 +760,252 @@ instance (i ~ G.C, GSerialiseProd f) => GSerialiseSum (G.M1 i c f) where
+----------------------------------------------------------------------------
+
+(.:) :: (c -> d) -> (a -> b -> c) -> (a -> b -> d)
+f .: g = \x y -> f (g x y)
+
+-- | Expressions describing the statically-computed size bounds on
+--   a type's possible values.
+type Size = Fix SizeF
+
+-- | The base functor for @Size@ expressions.
+data SizeF t
+    = AddF t t -- ^ Sum of two sizes.
+    | MulF t t -- ^ Product of two sizes.
+    | SubF t t -- ^ Difference of two sizes.
+    | AbsF t   -- ^ Absolute value of a size.
+    | NegF t   -- ^ Negation of a size.
+    | SgnF t   -- ^ Signum of a size.
+    | CasesF [Case t] -- ^ Case-selection for sizes. Used for sum types.
+    | ValueF Byte     -- ^ A constant value.
+    | ApF String (Byte -> Byte) t
+      -- ^ Application of a monotonic function to a size.
+    | forall a. Bi a => TodoF (forall x. Bi x => Proxy x -> Size) (Proxy a)
+      -- ^ A suspended size calculation ("thunk").
+  deriving Typeable
+
+instance Functor SizeF where
+    fmap f = \case
+        AddF x y  -> AddF (f x) (f y)
+        MulF x y  -> MulF (f x) (f y)
+        SubF x y  -> SubF (f x) (f y)
+        AbsF x    -> AbsF (f x)
+        NegF x    -> NegF (f x)
+        SgnF x    -> SgnF (f x)
+        CasesF xs -> CasesF (map (fmap f) xs)
+        ValueF x  -> ValueF x
+        ApF n g x -> ApF n g (f x)
+        TodoF g x -> TodoF g x
+
+instance Num (Fix SizeF) where
+    (+) = Fix .: AddF
+    (*) = Fix .: MulF
+    (-) = Fix .: SubF
+    negate = Fix . NegF
+    abs    = Fix . AbsF
+    signum = Fix . SgnF
+    fromInteger = Fix . ValueF . fromInteger
+
+instance Buildable t => Buildable (SizeF t) where
+    build x_ =
+        let showp2 c x y = bprint ("(" % build % " " % string % " " % build % ")") x c y
+        in case x_ of
+            AddF x y -> showp2 "+" x y
+            MulF x y -> showp2 "*" x y
+            SubF x y -> showp2 "-" x y
+            NegF x   -> bprint ("-" % build) x
+            AbsF x   -> bprint ("|" % build % "|") x
+            SgnF x   -> bprint ("sgn(" % build % ")") x
+            CasesF xs -> bprint ("{ " % build % "}") $ foldMap (bprint (build % " ")) xs
+            ValueF x  -> bprint shown (toInteger x)
+            ApF n _ x -> bprint (string % "(" % build % ")") n x
+            TodoF _ x -> bprint ("(_ :: " % shown % ")") (typeRep x)
+
+instance Buildable (Fix SizeF) where
+    build x = bprint build (unfix x)
+
+-- | Create a case expression from individual cases.
+szCases :: [Case Size] -> Size
+szCases = Fix . CasesF
+
+-- | An individual labeled case.
+data Case t = Case String t deriving (Typeable, Functor)
+
+-- | Discard the label on a case.
+caseValue :: Case t -> t
+caseValue (Case _ x) = x
+
+instance Buildable t => Buildable (Case t) where
+  build (Case n x) = bprint (string % "=" % build) n x
+
+-- | A range of values. Should satisfy the invariant @forall x. lo x <= hi x@.
+data Range b = Range { lo :: b, hi :: b }
+
+instance Num b => Num (Range b) where
```

Comment:
Does this instance really make sense? I'm not sure what it means to add, multiply and subtract ranges pointwise. Is that actually used?

Reply:
Yes, it is used in `szSimplify` to fold a `Size` up into a `Range Byte`. The `AddF` constructor for `Size` turns into `(+)` for `Range Byte`, and so on.

The various operations aren't defined pointwise, but as set operations. For example, `{ x | lo <= x <= hi } + `{ x' | lo' <= x' <= hi' }` is defined as `{ x + x' | lo <= x <= hi, lo' <= x' <= hi' }`. But it turns out that you can reproduce all of these calculations as calculations on the endpoints.

I think I should add an implementation for `abs` for completeness. As usual, I'm not sure if I can write a lawful `signum`.

Reply:
Initially this instance was written under the assumption that the range only contained non-negative values, but that really doesn't make sense given the implementation of `negate`. So I went ahead and implemented general interval arithmetic here.

Reply:
Hmmm I think that needs a comment clarifying that. For me if you say "as set operations", then I'd expect, say, "+" to work as set union (minimum on `lo`, maximum on `hi`). Not saying it doesn't make sense -- can pick any semantics we need of course. Just saying it needs some clarification in the comments :)